### PR TITLE
[ADD] business_requirement_sale: New module in 11.0

### PR DESCRIPTION
**Pull request by martintorresvm**
This module extends the functionality of Business Requirement Deliverable and to allow you to convert Requirement in new Sale Orders

Depends in 

- [x] https://github.com/OCA/business-requirement/pull/311

### CODE REVIEW

**Comment by martintorresvm** [comment]
cc @Tecnativa @pedrobaeza @sergio-teruel @Tardo

File: business_requirement_deliverable_sale/models/business.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+    @api.multi
+    @api.depends('sale_order_ids')
+    def _compute_order_count(self):
+        for order in self:
+            order.order_count = len(order.sale_order_ids)
+
+    @api.multi
+    def convert_requirement_sale(self):
+        import wdb
```
Comment (Tardo) [inline business_requirement_deliverable_sale/models/business.py]: :fire:

File: business_requirement_deliverable_sale/models/business.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        wdb.set_trace()
```
Comment (Tardo) [inline business_requirement_deliverable_sale/models/business.py]: :fire:

File: business_requirement_deliverable_sale/models/business.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        return action
+
+    @api.multi
+    def launch_convert_wizard(self):
+        vals = {
+            'br_id': self.id,
+            'brd_ids': [(6, 0, self.deliverable_lines.ids)],
+        }
+        wizard_obj = self.env['convert.requirement.sale']
+        # wizard = wizard_obj.with_context(br_ids=False).create(vals)
```
Comment (Tardo) [inline business_requirement_deliverable_sale/models/business.py]: If not used... :fire:

File: business_requirement_deliverable_sale/wizard/convert_requirement_sale.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        return values
+
+    def _prepare_sale_order_values(self, br_id):
+        if br_id:
+            values = {
+                'name': br_id.name,
+                'origin': br_id.name,
+                'partner_id': br_id.partner_id.id,
+            }
+        return values
```
Comment (Tardo) [inline business_requirement_deliverable_sale/wizard/convert_requirement_sale.py]: 'values' can be referenced before assignment

File: business_requirement_deliverable_sale/wizard/convert_requirement_sale.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+    def _create_sale_order(self, br_id):
+        vals = {}
```
Comment (Tardo) [inline business_requirement_deliverable_sale/wizard/convert_requirement_sale.py]: :fire:

File: business_requirement_deliverable_sale/wizard/convert_requirement_sale.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        if br_id:
+            vals = self._prepare_sale_order_values(br_id)
+            res = self.env['sale.order'].create(vals)
+            return res
+        return False
+
+    def _create_sale_order_line(self, br_id, so_id):
+        vals = {}
```
Comment (Tardo) [inline business_requirement_deliverable_sale/wizard/convert_requirement_sale.py]: :fire:

**Review by Tardo — CHANGES_REQUESTED** [review]

**Comment by martintorresvm** [comment]
Changes done @Tardo

**Comment by rafaelbn** [comment]
WIP until https://github.com/OCA/business-requirement/pull/311 is merged. It cannot be tested in runbot

**Comment by martintorresvm** [comment]
rebased @rafaelbn

File: business_requirement_deliverable_sale/models/business.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        """Get the right pricelist for product"""
+        uom_id = False
+        sale_price_unit = 0
+        product = self.product_id
+
+        if product:
+            uom_id = product.uom_id.id
+            sale_price_unit = product.list_price
+
+        sale_price_unit = self.product_id.list_price
```
Comment (Tardo) [inline business_requirement_deliverable_sale/models/business.py]: :fire:

File: business_requirement_deliverable_sale/models/business.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    @api.onchange('product_id')
+    def product_id_change(self):
+        """Get the right pricelist for product"""
+        uom_id = False
+        sale_price_unit = 0
+        product = self.product_id
+
+        if product:
+            uom_id = product.uom_id.id
+            sale_price_unit = product.list_price
```
Comment (Tardo) [inline business_requirement_deliverable_sale/models/business.py]: Here not is 'price'?

**Review by Tardo — CHANGES_REQUESTED** [review]

File: business_requirement_deliverable_sale/models/business.py (hunk 8 above)
Comment (martintorresvm) [reply business_requirement_deliverable_sale/models/business.py]: This is the same criteria as the product on `business.requirement.deliverable` model which for historic reason the default value decided is the product `list_price` if not exist a pricelist. But as the product `price` is computed by Odoo when is passed on the `with_context` function a pricelist so is suitable to be the sale_price_unit of the requirement

**Comment by martintorresvm** [comment]
changes done @Tardo

**Comment by Tardo** [comment]
Please, solve linter errors.

**Comment by martintorresvm** [comment]
I rebased

**Comment by pedrobaeza** [comment]
Why is this info at BR level? And required, which invalidates smart buttons?

Selección_016

**Comment by martintorresvm** [comment]
>Why is this info at BR level? And required, which invalidates smart buttons?

@pedrobaeza I extract this requirements from the meeting we had but probably need more detail or I misunderstood and I hope by asking for revision to find out what is left. 

Briefly I understand this:

We let user select one product and onchange place the standard price and quantity. Also have a checkbox `Detailed order` which means if checked we could edit `qty` and `sales price` as we wish.
Then if user try convert to quotation this product is used for the SO wizard and place any Deliverables as the `proto-sale order lines` then generate it
  
I made it required but probably should not and better hide button for convert to SO if false

**Comment by pedrobaeza** [comment]
Product is linked to each BRD, not at BR level. And on the wizard, you select which BRDs you want to include in the sales order, passing them as sale.order.line, using the product at BRD level and you decide one of this option:

* Pass quantity and the price unit.
* Put 1 as quantity and pass the subtotal as price unit.

**Comment by martintorresvm** [comment]
Changes done @pedrobaeza

**Comment by pedrobaeza** [comment]
Functional comments:

- Button must be called "Create Quotation", as we are not transforming the BR.
- On the shown wizard, the many2many should be put with widget `many2many_checkboxes` (or a similar name that present lines with checkboxes).
- The created quotation must preserve their numbering, not to use BR name.
- Navigate directly when accepting the wizard to the SO form view, instead of the list view.
- If sections are enabled, the quotation must put BRD sections as quotation sections.
- There's no smart-button in BR to navigate to already generated SOs (in list view mode).
- There's no many2one link in SO, which should be put in "Other information" page.
- Add in the BR chatter a mention that a SO has been created (even putting the embedded link in the message).
- Add the reversed one in the SO chatter, saying that it has been generated from BR etc.

Little technical one (as I haven't checked yet the code):

- I think we should rename the module to `business_requirement_sale`, as although it depends on BRD, the processes are for BR in general. What do you think?

**Comment by pedrobaeza** [comment]
Another technical one:

- Keep a link between sale.order.line and business.requirement.deliverable. Take this as a general rule: always that you generate something from other thing, you must add a link between them for traceability.

**Comment by martintorresvm** [comment]
@pedrobaeza All done but this:
>Add in the BR chatter a mention that a SO has been created (even putting the embedded link in the message).

This one not sure if could done easy, `trace_visibility` won't work over One2many fields. Hope I could develop a method soon for it

**Comment by martintorresvm** [comment]
@pedrobaeza About tracking one2many I found exist a module that should be soon on Odoo v11. Could we move forward and use it as soon this is merged https://github.com/OCA/social/pull/254 it's done?

**Comment by pedrobaeza** [comment]
I'm not talking about doing a tracking, but to manually post an entry through `message_post` method, and you can embed the link in it. Check

https://github.com/odoo/odoo/blob/5c157c57b0c56804576e61f49b431db82ee9fbb7/addons/sale_timesheet/models/sale_order.py#L248-L249

for an example.

**Comment by martintorresvm** [comment]
Changes done @pedrobaeza

File: business_requirement_deliverable_sale/__manifest__.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    "name": "Business Requirement Deliverable Sale",
+    'category': 'Business Requirements Management',
+    'summary': 'Convert Business Requirement into \
+                Sale Orders',
+    "version": "11.0.1.0.0",
+    "development_status": "Alpha",
+    "website": "https://github.com/OCA/business-requirement",
+    "author": "Tecnativa, "
+              "Odoo Community Association (OCA)",
+    "maintainers": ["your-github-login"],
```
Comment (pedrobaeza) [inline business_requirement_deliverable_sale/__manifest__.py]: This is incorrect. Put yourself or remove it.

File: business_requirement_deliverable_sale/models/business.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    )
+    detailed_order = fields.Boolean(
+        help='Check this if need generated a detailed order'
+    )
+
+    @api.multi
+    @api.depends('sale_order_ids')
+    def _compute_order_count(self):
+        for order in self:
+            order.order_count = len(order.sale_order_ids)
```
Comment (pedrobaeza) [inline business_requirement_deliverable_sale/models/business.py]: Use `read_group` for performance. Example:

https://github.com/OCA/account-financial-tools/blob/55639a62200cd5001774d87586e4c473918b1464/account_credit_control/models/credit_control_run.py#L69-L78

File: business_requirement_deliverable_sale/models/business.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+                }}
+
+
+class BusinessRequirementDeliverable(models.Model):
+    _inherit = 'business.requirement.deliverable'
+
+    sale_order_line_id = fields.Many2one(
+        comodel_name='sale.order.line',
+        string='Order Line',
+        track_visibility='onchange',)
```
Comment (pedrobaeza) [inline business_requirement_deliverable_sale/models/business.py]: Why adding `track_visibility` here? It's not needed.

File: business_requirement_deliverable_sale/models/sale_order.py [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+    business_requirement_id = fields.Many2one(
+        comodel_name='business.requirement',
+        string='Business requirement',
+    )
+
+
+class SaleOrderLine(models.Model):
+    _inherit = 'sale.order.line'
+
+    business_requirement_deliverable_section_id = fields.Many2one(
```
Comment (pedrobaeza) [inline business_requirement_deliverable_sale/models/sale_order.py]: No need of this. What you have to do is to create sales sections, and link them with BR section, and also put the sale order lines with the corresponding sale sections.

File: business_requirement_deliverable_sale/readme/DESCRIPTION.rst [hunk 13]
```diff
@@ -0,0 +1,2 @@
+This module extends the functionality of Business Requirement Deliverable
+and to allow you to convert Requirement in new Sale Orders
```
Comment (pedrobaeza) [inline business_requirement_deliverable_sale/readme/DESCRIPTION.rst]: This is not exactly accurate. BR = sales order. BRD = sales order lines.

Comment (pedrobaeza) [reply business_requirement_deliverable_sale/readme/DESCRIPTION.rst]: And please add usage section with the wizard explanation.

File: business_requirement_deliverable_sale/wizard/convert_requirement_sale.py [hunk 14]
```diff
@@ -0,0 +1,135 @@
+# Copyright 2019 Tecnativa Victor M.M. Torres>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import api, models, fields, _
+
+
+class ConvertRequirementSale(models.TransientModel):
+    """Wizard to guide create new SO from requirement"""
+
+    _name = 'convert.requirement.sale'
```
Comment (pedrobaeza) [inline business_requirement_deliverable_sale/wizard/convert_requirement_sale.py]: Call it with the same prefix as the rest: `business.requirement.convert.sale` or similar.

File: business_requirement_deliverable_sale/wizard/convert_requirement_sale.py [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+                'active_model'
+        ) == 'business.requirement' and self._context.get('active_id'):
+            res.update({
+                'br_id': self._context['active_id']
+            })
+        if 'brd_ids' in fields and not res.get('brd_ids') and res.get('br_id'):
+            res.update({'brd_ids': [(6, 0, self.env[
+                'business.requirement'].browse(
+                    res['br_id']).deliverable_lines.ids)]})
+
```
Comment (pedrobaeza) [inline business_requirement_deliverable_sale/wizard/convert_requirement_sale.py]: Don't let empty lines in the middle of a method.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: business_requirement_deliverable_sale/__manifest__.py [hunk 16]
```diff
@@ -0,0 +1,26 @@
+# Copyright 2019 Tecnativa Victor M.M. Torres>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+{
+    "name": "Business Requirement Deliverable Sale",
+    'category': 'Business Requirements Management',
```
Comment (sergio-teruel) [inline business_requirement_deliverable_sale/__manifest__.py]: Use simple or double cuotes, as you prefer, but always the same style..

File: business_requirement_deliverable_sale/models/__init__.py [hunk 17]
```diff
@@ -0,0 +1,5 @@
+# Copyright 2019 Tecnativa Victor M.M. Torres>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from . import business
```
Comment (sergio-teruel) [inline business_requirement_deliverable_sale/models/__init__.py]: I think is better call the file business_requirement

File: business_requirement_deliverable_sale/models/business.py (hunk 10 above)
Comment (sergio-teruel) [reply business_requirement_deliverable_sale/models/business.py]: Would be better use a search_count in this case??
count = self.env['sale.order'].search_count([('business_requirement_id', '=', br.id)])

File: business_requirement_deliverable_sale/models/business.py [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+        compute='_compute_order_count',
+    )
+    detailed_order = fields.Boolean(
+        help='Check this if need generated a detailed order'
+    )
+
+    @api.multi
+    @api.depends('sale_order_ids')
+    def _compute_order_count(self):
+        for order in self:
```
Comment (sergio-teruel) [inline business_requirement_deliverable_sale/models/business.py]: for business_requirement in self: or for br in self:

File: business_requirement_deliverable_sale/models/business.py [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+            order.order_count = len(order.sale_order_ids)
+
+    @api.multi
+    def convert_requirement_sale(self):
+        vals = {
+            'br_id': self.id,
+            'brd_ids': [(6, 0, self.deliverable_lines.ids)]
```
Comment (sergio-teruel) [inline business_requirement_deliverable_sale/models/business.py]: I think there are many abbreviations

File: business_requirement_deliverable_sale/models/business.py [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+        }
+        wizard_obj = self.env['convert.requirement.sale']
+        wizard = wizard_obj.create(vals)
```
Comment (sergio-teruel) [inline business_requirement_deliverable_sale/models/business.py]: You can use directly
wiz = self.env['convert.requirement.sale'].create(vals)

File: business_requirement_deliverable_sale/models/business.py [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+            'br_id': self.id,
+            'brd_ids': [(6, 0, self.deliverable_lines.ids)],
+        }
+        wizard_obj = self.env['convert.requirement.sale']
+        wizard = wizard_obj.create(vals)
+        action = wizard.wizard_view()
+        return action
+
+    @api.multi
+    def open_orders(self):
```
Comment (sergio-teruel) [inline business_requirement_deliverable_sale/models/business.py]: See this example
https://github.com/OCA/stock-logistics-workflow/blob/d87ee07a99f8d90ec74713e2c7e83aa6e7f55b90/stock_batch_picking/models/stock_batch_picking.py#L212

**Review by sergio-teruel — CHANGES_REQUESTED** [review]

File: business_requirement_deliverable_sale/models/business.py (hunk 10 above)
Comment (pedrobaeza) [reply business_requirement_deliverable_sale/models/business.py]: `read_group` is more performant for several records, and more or less the same as `search_count` for only one.

**Comment by martintorresvm** [comment]
Changes done @pedrobaeza

**Comment by martintorresvm** [comment]
Changes done @pedrobaeza

File: business_requirement_sale/models/business.py [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+            'brd_ids': [(6, 0, self.deliverable_lines.ids)],
+        }
+        wizard_obj = self.env['business.convert.requirement.sale']
+        wizard = wizard_obj.create(vals)
+        action = wizard.wizard_view()
+        return action
+
+    @api.multi
+    def open_orders(self):
+        for self in self:
```
Comment (Tardo) [inline business_requirement_sale/models/business.py]:
```suggestion
        for rec in self:
```

File: business_requirement_sale/models/business.py [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+            domain = [('business_requirement_id', '=', self.id)]
```
Comment (Tardo) [inline business_requirement_sale/models/business.py]:
```suggestion
            domain = [('business_requirement_id', '=', rec.id)]
```

File: business_requirement_sale/models/business.py [hunk 24]
```diff
@@ -0,0 +1,104 @@
...
+            br_id = 0
+            if self.state in ('draft', 'confirmed'):
+                br_id = self.id
```
Comment (Tardo) [inline business_requirement_sale/models/business.py]:
```suggestion
                br_id = rec.id
```

File: business_requirement_sale/models/business.py [hunk 25]
```diff
@@ -0,0 +1,104 @@
...
+            groupby=['business_requirement_id'],
+        )
+        if fetch_data:
+            aux_count = fetch_data[0].get(
+                'business_requirement_id_count', 0)
+        for rec in self:
+            rec.order_count = aux_count
+
+    @api.multi
+    def convert_requirement_sale(self):
```
Comment (Tardo) [inline business_requirement_sale/models/business.py]: self.ensure_one()?

File: business_requirement_sale/models/business.py [hunk 26]
```diff
@@ -0,0 +1,104 @@
...
+            'br_id': self.id,
+            'brd_ids': [(6, 0, self.deliverable_lines.ids)]
+        }
+        wizard_obj = self.env['business.convert.requirement.sale']
+        wizard = wizard_obj.create(vals)
+        action = wizard.wizard_view()
+        return action
+
+    @api.multi
+    def launch_convert_wizard(self):
```
Comment (Tardo) [inline business_requirement_sale/models/business.py]: self.ensure_one()?

File: business_requirement_sale/wizard/convert_requirement_sale.py [hunk 27]
```diff
@@ -0,0 +1,159 @@
...
+        if br_id and so_id and brd:
+            values = {
+                'name': br_id.name,
+                'order_id': so_id.id,
+                'product_id': brd.product_id.id,
+                'product_uom_qty': brd.qty,
+                'price_unit': brd.sale_price_unit,
+                'business_requirement_deliverable_id': brd.id,
+            }
+            if values:
```
Comment (Tardo) [inline business_requirement_sale/wizard/convert_requirement_sale.py]: :fire:

File: business_requirement_sale/wizard/convert_requirement_sale.py [hunk 28]
```diff
@@ -0,0 +1,159 @@
...
+                                'line_section_id': line_section.id
+                            })
+                    if sol and self.totaled_method == 'totaled':
+                        sol._totaled_method()
+                    brd.write({
+                        'sale_order_line_id': sol.id,
+                    })
+
+    @api.multi
+    def apply(self):
```
Comment (Tardo) [inline business_requirement_sale/wizard/convert_requirement_sale.py]: self.ensure_one()

File: business_requirement_sale/wizard/convert_requirement_sale.py [hunk 29]
```diff
@@ -0,0 +1,159 @@
...
+        """Create SO and get to right view"""
+        so_id = self._create_sale_order(self.br_id)
+        brd_ids = self.brd_ids
```
Comment (Tardo) [inline business_requirement_sale/wizard/convert_requirement_sale.py]: :fire:

File: business_requirement_sale/wizard/convert_requirement_sale.py [hunk 30]
```diff
@@ -0,0 +1,159 @@
...
+        if so_id:
+            self._create_sale_order_line(self.br_id, so_id, brd_ids)
```
Comment (Tardo) [inline business_requirement_sale/wizard/convert_requirement_sale.py]:
```suggestion
            self._create_sale_order_line(self.br_id, so_id, self.brd_ids)
```

**Review by Tardo — CHANGES_REQUESTED** [review]

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**